app/reflexes/admin/orders_reflex.rb

```diff
@@ -8,8 +8,13 @@ def capture
       payment_capture = OrderCaptureService.new(@order)
 
       if payment_capture.call
-        morph dom_id(@order), render(partial: "spree/admin/orders/table_row",
-                                     locals: { order: @order.reload, success: true })
+        cable_ready.inner_html(
+          selector: dom_id(@order),
+          html: render(partial: "spree/admin/orders/table_row",
+                       locals: { order: @order.reload, success: true })
+        ).broadcast
+
+        morph :nothing
```

Comment:
That's an interesting observation. It sounds like a bug in StimulusReflex to me. I don't like this new pattern but if it works then it's good.

I tried to find any information about why this is happening. And I found an interesting section in the docs:

> `inner_html` completely wipes out any Stimulus controllers present in the replaced DOM hierarchy and doesn't respect the data-reflex-permanent attribute. How can we adapt this so that both morph operations are performed by the morphdom library?
> https://docs.stimulusreflex.com/guide/morph-modes.html#real-world-example-pagy-refactoring

So there's a difference between inner_html and morphdom and morphdom seems to keep as much as possible, therefore not disconnecting and reconnecting controllers. I think it tries to update the existing tooltip elements without replacing them.

Maybe a problem of the tooltip controller is that it renders the tooltip on connect and is not aware of data attribute changes after that.

Another thing I found is: 

> You cannot change the attributes of your morph target.
> https://docs.stimulusreflex.com/appendices/troubleshooting.html#you-cannot-change-the-attributes-of-your-morph-target

So either solution means that the `tr` attribute for order state never updates. This is probably a pre-existing bug. But looking at [morphdom](https://github.com/patrick-steele-idem/morphdom) it seems to be able to update attributes on the target. So, I'm not sure where that limitation comes from or if the warning is only there in case StimulusReflex decides to replace inner_html only (which it seems to do in some cases).

Maybe we should test if the state change is a problem and then we need to use outer_html. Otherwise we are good.

Reply:
I think this describe quite neatly the issue and the recommended solution here : 
https://github.com/stimulusreflex/stimulus_reflex/issues/314#issuecomment-702479357

I am wondering if we should rewrite the tooltip stimulus controller to use `stimulus-reflex:after` callback as described above, it seems to me to be the better approach. It would prevent the issue in any other page that uses the tooltip controller.As it stands, the onus is on the reflex to do the correct render, and I feel like it's not its job to have to worry about any stimulus controller it's not interacting directly with. 

That said the stimulus docs do describe the pattern used here : https://docs.stimulusreflex.com/guide/cableready.html#using-cableready-inside-a-reflex-action 

What do you guys think ?

Reply:
Thank you for the links, Gaetan, very interesting. I didn't know that you could also queue cable_ready events without broadcasting them immediately and that they get then broadcasted after the StimulusReflex morph.

I guess the idealistic solution for the tooltip is to render the tooltip server-side and include it in the morphing partial. Hm, that's actually already done in other places. Do we just need to get rid of `insertToolTipMarkup` and replace it with a partial?

Reply:
I believe original intend was to use a partial : https://github.com/openfoodfoundation/openfoodnetwork/commit/b76d9045ce733c61c32919cbf51eeae37c2c48fd but then Matt did some refactoring here : https://github.com/openfoodfoundation/openfoodnetwork/commit/d64a9afece84105e96e591bb935726ff489bff6c which led to the current behaviour.

I had a go at using `stimulus-reflex` callbacks, I can trigger a re render but for some reason it's not finding the data attributes linked to the element with the tooltip controller :confused:  

From the linked I shared above :  
> In the context of SR (Stimulus Reflex), the expectation is that you would specify the UI you want to exist in the browser on the server. There's  no need for a render stage because the markup is already perfect before it is sent down the wire. There is no build stage, no rehydration. We're not taking over the task of redrawing the UI from the browser.

I'll look at moving back to a partial.